app/src/keymap.c

```diff
@@ -685,146 +686,83 @@ int zmk_keymap_reset_settings(void) { return -ENOTSUP; }
 
 #endif // IS_ENABLED(CONFIG_ZMK_KEYMAP_SETTINGS_STORAGE)
 
-int zmk_keymap_apply_position_state(uint8_t source, zmk_keymap_layer_id_t layer_id,
-                                    uint32_t position, bool pressed, int64_t timestamp) {
-    const struct zmk_behavior_binding *binding =
-        zmk_keymap_get_layer_binding_at_idx(layer_id, position);
-    struct zmk_behavior_binding_event event = {
-        .layer = layer_id,
-        .position = position,
-        .timestamp = timestamp,
-#if IS_ENABLED(CONFIG_ZMK_SPLIT)
-        .source = source,
-#endif
-    };
-
-    LOG_DBG("layer_id: %d position: %d, binding name: %s", layer_id, position,
-            binding->behavior_dev);
-
-    return zmk_behavior_invoke_binding(binding, event, pressed);
-}
-
-int zmk_keymap_position_state_changed(uint8_t source, uint32_t position, bool pressed,
-                                      int64_t timestamp) {
-    if (pressed) {
-        zmk_keymap_active_behavior_layer[position] = _zmk_keymap_layer_state;
-    }
-
+int zmk_keymap_raise_binding_event_at_layer_idx(zmk_keymap_layer_id_t layer_id, uint8_t source,
+                                                uint32_t position, enum trigger_type type,
+                                                int64_t timestamp) {
     // We use int here to be sure we don't loop layer_idx back to UINT8_MAX
-    for (int layer_idx = ZMK_KEYMAP_LAYERS_LEN - 1;
-         layer_idx >= LAYER_ID_TO_INDEX(_zmk_keymap_layer_default); layer_idx--) {
-        zmk_keymap_layer_id_t layer_id = LAYER_INDEX_TO_ID(layer_idx);
+    for (int layer_idx = layer_id; layer_idx >= LAYER_ID_TO_INDEX(_zmk_keymap_layer_default);
```

Comment:
Related to the other comment on ID vs. index, I believe this one should be looping over indices, not IDs, like before.